 design/README.MD

```diff
@@ -0,0 +1,456 @@
+It will also create a shim
+that will convert the Dafny to native
+and then call the native implementation.
+
+```mermaid
+---
+title: Wrapped Library
+---
+flowchart RL
+
+subgraph Wrapping[Wrapping Library]
+    direction RL
+    WrappingInterface[Generated Dafny SDK Interface] --> ToNativeW[Convert input \nto native]
+    ToDafnyW[Convert output \nto Dafny] --> WrappingInterface
+end
+
+subgraph Library[Native Library]
+    direction RL
+    Interface[Public Interface] --> ToDafny[Convert input \nto Dafny]
+    ToDafny --> DafnyImplementation[Dafny implementation]
+    DafnyImplementation --> ToNative[Convert output \nto native]
+    ToNative --> Interface
+end
+
+ToNativeW --> Interface
+Interface --> ToDafnyW
+
+DafnyTest --> WrappingInterface
+WrappingInterface --> DafnyTest
+
+```
+
+The goal is for Dafny developers
+to approximate native calls into Dafny libraries.
+`smithy-dafny` uses the wrapped library feature
+to test the correctness of polymorph-runtime features.
+
+Like this: `var output := ToDafny(ToNative(Work(ToDafny(ToNative(input)))))`
+
+## Polymorph development
+
+When developing `polymorph-X`
+the first thing to ask is:
+Does `smithy-X` exist?
+
+If it does exist then:
+`polymorph-X` SHOULD re-use as much `smithy-X` code-gen as possible.
+`polymorph-X` MUST use the same native types as `smithy-X`.
+
+If it does not exists then look at the AWS SDK in X:
+`polymorph-X` SHOULD use the same native types as the native AWS SDK in X runtime.
+
+Regardless `polymorph-X` MUST convert these native types
+to the Dafny types produced by `smithy-dafny`.
+
+```mermaid
+---
+title: Generated Code
+---
+flowchart LR
+
+subgraph Project
+    NativeInterface
+    NativeTypes
+    ToNative
+    ToDafny
+    DafnyTypes
+    DafnyInterface
+end
+
+subgraph Polymorph
+    SmithyNative
+end
+
+Model --> Polymorph
+Polymorph --> NativeInterface
+SmithyNative --> NativeTypes
+Polymorph --> ToNative
+Polymorph --> ToDafny
+Model --> Dafny
+Dafny --> DafnyTypes
+Dafny --> DafnyInterface
+```
+
+In this way the dual inputs and output
+of the type conversion dual functions are defined.
+
+```mermaid
+---
+title: Type conversion dual
+---
+flowchart LR
+NativeTypes --> ToDafny --> DafnyTypes
+DafnyTypes --> ToNative --> NativeTypes
+```
+
+However there may be many to one relationships.
+Perhaps `string` and `char[]`.
+These two would convert to `ValidUTF8Bytes` or `seq<uint8>`.
+
+Also reference types may not be exactly equal.
+While a native array would convert to a `seq`
+when converting back the new array
+will have the same elements in the same order
+but may not be exactly the same reference object.
+We can refer to this as `SemanticallyEqual`.
+
+The spirit of this can be expressed
+```dafny
+lemma ConversionsCorrectlyInvert<D, N>(ToDafny<D,N>, ToNative<N, D>, SemanticallyEqual)
+  ensures forall d, n ::
+    SemanticallyEqual(ToNative(d), n) <==> SemanticallyEqual(ToDafny(n), d)
+```
```

Comment:
FWIW if your `N` is a Dafny extern declaration for the native type, this could at least actually type check even if it's obviously not provable. When we have Dafny standard libraries for interop and `ToDafny` and `ToNative` are actually Dafny functions, you could imagine somehow generating tests from this specification. :)